### PR TITLE
iptables: use `ISTIO_OUTPUT` chain for DNS UDP rules

### LINKED ISSUE #50532: sidecar iptables - istio-specific rules should only be inserted into custom chains
**Describe the feature request**

In ambient, we create custom `ISTIO_XXX` prefixed chains for our custom iptables rules, and only put our custom rules in these prefixed chains.

This makes cleanup and upgrade simpler (main table/chain has one jump to our custom chain, we can remove the drop and the prefixed chains safely), and is the only pattern we should follow for Istio-specific iptables rules, no matter the context (sidecar or node).

In sidecar, we are still slamming a bunch of rules into the main table/chain, e.g: many of the sidecar DNS rules are just in OUTPUT and should be in a custom chain named ISTIO_OUTPUT or similar: https://github.com/istio/istio/blob/5789705232cfcb5c69a48631da4ce5f689b475ff/tools/istio-iptables/pkg/capture/testdata/dns-uid-gid.golden#L24

Sidecar rule creation should be tweaked to create and insert all rules into custom chains, like ambient does.

We have added a WARN nag to rules cleanup that will flag any non-jump rules in main tables: https://github.com/istio/istio/pull/50328#discussion_r1567224756 - a PR to fix this issue should eliminate all outstanding WARNs.

**Describe alternatives you've considered**

**Affected product area (please put an X in all that apply)**

[ ] Ambient
[ ] Docs
[x] Dual Stack
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Affected features (please put an X in all that apply)**

[ ] Multi Cluster
[ ] Virtual Machine
[ ] Multi Control Plane

**Additional context**

### PR DESCRIPTION
**Pull request by eoinfennessy**
**Please provide a description of this PR:**

Fixes #50532

### CODE REVIEW

**Review by costinm — COMMENTED** [review]
I think it's a good change but pretty risky - can you confirm the result by getting an iptables-save on a real pod ( the golden files are pretty fake - not sure why we use uid-owner 3 and 4, not something we'll have in practice )  ? 


(long term): If you are interested in iptables and capture - I would save the 2 in a doc, perhaps one capture from a pod with all iptables-related annotations - and evaluate the differences and if we can merge the behavior. It is not clear at all why we need 2 different iptables code paths both operating on the pod - in the past ambient was in host namespace and it made sense.

**Comment by costinm** [comment]
BTW - last part of my comment is not related to this PR, just suggestions if you want to explore iptables further, it's certainly a large project...

For this PR - just the part on checking the duplicated add.

**Comment by eoinfennessy** [comment]
Thanks @costinm!

> Golden files don't look right - ISTIO_OUTPUT is created 2x.

ISTIO_OUTPUT is created once in the `nat` table and once in the `raw` table (is this required because some of the rules created are added to the `raw` table?):
```
iptables -t nat -N ISTIO_OUTPUT
iptables -t raw -N ISTIO_OUTPUT
...
ip6tables -t nat -N ISTIO_OUTPUT
ip6tables -t raw -N ISTIO_OUTPUT
```

> Also there is no jump from OUTPUT to ISTIO_OUTPUT showing up - maybe I missed it or was already there for the previous ISTIO_OUTPUT rule ? I don't remember how the jump is added - please check it doesn't have "-p tcp".

This jump is currently being created with `-p tcp`:
```
iptables -t nat -A OUTPUT -p tcp -j ISTIO_OUTPUT
```
Do we want to now create this without the `-p tcp` argument? Or create another jump with `-p udp`? I think I'm missing the jumps to the `raw` table too.

> can you confirm the result by getting an iptables-save on a real pod ( the golden files are pretty fake - not sure why we use uid-owner 3 and 4, not something we'll have in practice ) - and compare it against what ambient does ?

Yes, I'll do some comparisons with between this and the current behavior for both sidecar and ambient.

**Review by bleggett — COMMENTED** [review]
This LGTM conceptually and is a pretty straightforward rename.

We have decent testing around this, so if/when the tests pass, I am good. TY @eoinfennessy !

File: tools/istio-iptables/pkg/capture/run.go [hunk 1]
```diff
@@ -616,7 +616,7 @@ func HandleDNSUDP(
 		ext:      ext,
 		ops:      ops,
 		table:    constants.NAT,
-		chain:    constants.OUTPUT,
```
Comment (bleggett) [inline tools/istio-iptables/pkg/capture/run.go:619]: You need to insert the jump to the istio chain, since we are no longer using the default one.

Comment (eoinfennessy) [reply tools/istio-iptables/pkg/capture/run.go:619]: Thank you, @bleggett! I've added the required jumps in the latest commit. The integration tests look OK now.

**Review by hzxuzhonghu — COMMENTED** [review]
with this one rule 
-A OUTPUT -p udp -j ISTIO_OUTPUT

I am not sure i can get what really it fix here

**Comment by eoinfennessy** [comment]
> with this one rule 
> -A OUTPUT -p udp -j ISTIO_OUTPUT
> 
> I am not sure i can get what really it fix here

Hi @hzxuzhonghu, this jump is required because all UDP rules have been moved from the main OUTPUT chain to the ISTIO_OUTPUT chain.

**Review by bleggett — APPROVED** [review]
LGTM! ty @eoinfennessy!

**Comment by howardjohn** [comment]
> > with this one rule
> > -A OUTPUT -p udp -j ISTIO_OUTPUT
> > I am not sure i can get what really it fix here
> 
> Hi @hzxuzhonghu, this jump is required because all UDP rules have been moved from the main OUTPUT chain to the ISTIO_OUTPUT chain.

Shouldn't we conditionally jump? Like only for dport 53/sport 15053?

**Comment by bleggett** [comment]
> > > with this one rule
> > > -A OUTPUT -p udp -j ISTIO_OUTPUT
> > > I am not sure i can get what really it fix here
> > 
> > 
> > Hi @hzxuzhonghu, this jump is required because all UDP rules have been moved from the main OUTPUT chain to the ISTIO_OUTPUT chain.
> 
> Shouldn't we conditionally jump? Like only for dport 53/sport 15053?

The existing rules only apply if the traffic meets those reqs (is udp/ports, etc) so it amounts to the same thing. 

Putting a rule in a custom chain and jumping to it doesn't change rule evaluation over how they would work in the primary/default chain (where they were before)

**Pull request merged**